src/Streamly/Internal/Data/Fold.hs

```diff
@@ -1445,40 +1407,45 @@ demuxDefaultWith f kv (Fold dstep dinitial dextract) =
                         if Map.size runMap1 == 0
                         then fDone doneMap1
                         else Partial $ fPartial doneMap1 runMap1
-            Done _ -> error "Bug: demuxDefaultWith: Done fold"
+            Done _ -> error "Bug: demuxWith: Done fold"
 
     step (DemuxMapAndDefault dacc doneMap runMap) a = do
-        let (k, a1) = f a
+        let k = f a
+            finished = Map.lookup k doneMap
         case Map.lookup k runMap of
-            Nothing -> do
-                res <- dstep dacc (k, a1)
-                return
-                    $ Partial
-                    $ case res of
-                          Partial s -> DemuxMapAndDefault s doneMap runMap
-                          Done b -> DemuxOnlyMap b doneMap runMap
-            Just fld ->
-                runFold
+            Nothing
+                | isNothing finished ->
+                    runFold
+                    (DemuxMapAndDefault dacc)
+                    (Partial . DemuxOnlyDefault dacc)
+                    doneMap runMap fd k a
```

Comment:
This has become ugly. 1) Although all folds in runMap are done we are not directly going to DemuxOnlyDefault state, we are running the default fold in this state only 2) only if the default fold returns Done then we will go to DemuxOnlyDefault State.